src/Data/TypeRepMap/Internal.hs

```diff
@@ -390,14 +387,19 @@ fromTriples kvs = TypeRepMap (GHC.fromList fpAs) (GHC.fromList fpBs) (GHC.fromLi
 ----------------------------------------------------------------------------
 
 fromSortedList :: forall a . [a] -> [a]
-fromSortedList l = IM.elems $ fst $ go 0 0 mempty (IM.fromList $ zip [0..] l)
+fromSortedList l = runST $ do
+    let n = length l
+    let arrOrigin = fromList l
+    arrResult <- thawArray arrOrigin 0 n
+    _ <- go n 0 0 arrResult arrOrigin
+    toList <$> freezeArray arrResult 0 n
```

Comment:
Looks like you can use `unsafeFreezeArray` here:

* https://hackage.haskell.org/package/primitive-0.6.4.0/docs/Data-Primitive-Array.html#v:unsafeFreezeArray

Both `thawArray` and `freezeArray` functions create a copy of an array by cloning, as I can see from the documentation. So in most cases one of the two (`freeze` or `thaw`) can be `unsafe` to make implementation even more efficient. In our case we can't use `unsafeThaw` because we actually use `arrOrigin` after `thaw`.

Reply:
yes, same goes for the `adjust` patch (I wanted to comment but it was merged already)